### PR TITLE
Router request / response debug logging

### LINKED ISSUE #2360: Wire logs with request/response body
### Describe your problem(s)
logs with request/response body when it is enabled.

### Describe your solution

Add descriptive debug and trace logs.
https://www.envoyproxy.io/docs/envoy/latest/operations/cli#cmdoption-component-log-level

### How will you implement it

---

### PR DESCRIPTION
**Pull request by Amila-Rukshan**
### Purpose
Support debug logging in router. Two Lua filters added before and after the Lua filter for interceptors. To view the debug logs in router following should be enabled. Each logging section logs set of headers and the body if available.

1. Set trailing args with for component level debug log for lua.
```
TRAILING_ARGS=--component-log-level http:debug,http2:debug,conn_handler:debug,lua:debug
```
2. Add following config in `log_config.toml` file.
```
[routerDebugLog]
enable = true
```
Following is a sample output log when the debug log is enabled as mentioned above.
```
[2022-04-11 13:29:19.539][14][debug][lua] [source/extensions/filters/http/lua/lua_filter.cc:231] resuming body due to end stream
[2022-04-11 13:29:19.539][14][debug][lua] [source/extensions/filters/http/lua/lua_filter.cc:792] script log: 
>> request before interceptors >> :authority: localhost:9095
>> request before interceptors >> :path: /book-store/v1/books
>> request before interceptors >> :method: POST
>> request before interceptors >> :scheme: https
>> request before interceptors >> user-agent: curl/7.68.0
>> request before interceptors >> accept: */*
>> request before interceptors >> authorization: Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiIsIng1dCI6Ik1UWm1ZemsxWlRJeVkyTmlNR00wT1RJMFpqTXdNRE5rTldZek16QTJOVEl4TWpRelpERTFNdz09In0.eyJhdWQiOiJodHRwOlwvXC9vcmcud3NvMi5hcGltZ3RcL2dhdGV3YXkiLCJzdWIiOiJhZG1pbiIsInNjb3BlIjoicmVhZDpwZXRzIiwiaXNzIjoiaHR0cHM6XC9cL2xvY2FsaG9zdDo5MDk1XC90ZXN0a2V5Iiwia2V5dHlwZSI6IlBST0RVQ1RJT04iLCJleHAiOjE2NDk2ODczNTcsImlhdCI6MTY0OTY4Mzc1NywianRpIjoiZDkzNzQ4NDEtNmVkYi00Zjc5LTgzMWQtOWVmOTBiMDMxNGUzIn0.nN0dePrkPGBaiHK34KXeb1hHTAflSMmqDTKZbXfWSoVUXH17bF5KtEuAunwuS8pO84aga3FlkClbH6W8W21KLmeOtccSAFl2lIqTzaMMD_VgwuXw99Pwj2C2qAelnROJxSG35tpJQTtd960ZY6DxUp6XlYlu-n_3oNrpLpTfYMnDV3h6i0M5P2L4wb6P8BV-xnDTbMZrEZRb7O4k26aeJvGuByFwzy4SGxHPSa4tHY-iIvo8wPMAJhCqebo_cy2Edjs-KvEi5KbRhB3WmADv8Dfm7BzxIN4xv_Miulo372tIGDyXdeL3t-2aE9CxeAR9Ruu_ovPyUhEOtCYmOwdSLQ
>> request before interceptors >> content-type: application/json
>> request before interceptors >> content-length: 23
>> request before interceptors >> x-forwarded-proto: https
>> request before interceptors >> x-request-id: 57832b70-446b-9945-9d63-1ceb3be51734
{"name":"The Prisoner"}

[2022-04-11 13:29:19.539][14][debug][lua] [source/extensions/filters/common/lua/lua.cc:39] coroutine finished
[2022-04-11 13:29:19.539][14][debug][lua] [source/extensions/filters/common/lua/lua.cc:39] coroutine finished
[2022-04-11 13:29:19.539][14][debug][lua] [source/extensions/filters/http/lua/lua_filter.cc:470] yielding for full body
[2022-04-11 13:29:19.539][14][debug][lua] [source/extensions/filters/common/lua/lua.cc:42] coroutine yielded
[2022-04-11 13:29:19.539][14][debug][lua] [source/extensions/filters/http/lua/lua_filter.cc:231] resuming body due to end stream
[2022-04-11 13:29:19.539][14][debug][lua] [source/extensions/filters/http/lua/lua_filter.cc:792] script log: 
>> request after interceptors >> :authority: localhost:9095
>> request after interceptors >> :path: /book-store/v1/books
>> request after interceptors >> :method: POST
>> request after interceptors >> :scheme: https
>> request after interceptors >> user-agent: curl/7.68.0
>> request after interceptors >> accept: */*
>> request after interceptors >> authorization: Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiIsIng1dCI6Ik1UWm1ZemsxWlRJeVkyTmlNR00wT1RJMFpqTXdNRE5rTldZek16QTJOVEl4TWpRelpERTFNdz09In0.eyJhdWQiOiJodHRwOlwvXC9vcmcud3NvMi5hcGltZ3RcL2dhdGV3YXkiLCJzdWIiOiJhZG1pbiIsInNjb3BlIjoicmVhZDpwZXRzIiwiaXNzIjoiaHR0cHM6XC9cL2xvY2FsaG9zdDo5MDk1XC90ZXN0a2V5Iiwia2V5dHlwZSI6IlBST0RVQ1RJT04iLCJleHAiOjE2NDk2ODczNTcsImlhdCI6MTY0OTY4Mzc1NywianRpIjoiZDkzNzQ4NDEtNmVkYi00Zjc5LTgzMWQtOWVmOTBiMDMxNGUzIn0.nN0dePrkPGBaiHK34KXeb1hHTAflSMmqDTKZbXfWSoVUXH17bF5KtEuAunwuS8pO84aga3FlkClbH6W8W21KLmeOtccSAFl2lIqTzaMMD_VgwuXw99Pwj2C2qAelnROJxSG35tpJQTtd960ZY6DxUp6XlYlu-n_3oNrpLpTfYMnDV3h6i0M5P2L4wb6P8BV-xnDTbMZrEZRb7O4k26aeJvGuByFwzy4SGxHPSa4tHY-iIvo8wPMAJhCqebo_cy2Edjs-KvEi5KbRhB3WmADv8Dfm7BzxIN4xv_Miulo372tIGDyXdeL3t-2aE9CxeAR9Ruu_ovPyUhEOtCYmOwdSLQ
>> request after interceptors >> content-type: application/json
>> request after interceptors >> content-length: 23
>> request after interceptors >> x-forwarded-proto: https
>> request after interceptors >> x-request-id: 57832b70-446b-9945-9d63-1ceb3be51734
{"name":"The Prisoner"}

[2022-04-11 13:29:19.539][14][debug][lua] [source/extensions/filters/common/lua/lua.cc:39] coroutine finished
[2022-04-11 13:29:19.539][14][debug][http] [source/common/http/filter_manager.cc:953] [C12][S11928103386873881598] Sending local reply with details route_not_found
[2022-04-11 13:29:19.539][14][debug][lua] [source/extensions/filters/http/lua/lua_filter.cc:470] yielding for full body
[2022-04-11 13:29:19.539][14][debug][lua] [source/extensions/filters/common/lua/lua.cc:42] coroutine yielded
[2022-04-11 13:29:19.539][14][debug][lua] [source/extensions/filters/http/lua/lua_filter.cc:231] resuming body due to end stream
[2022-04-11 13:29:19.539][14][debug][lua] [source/extensions/filters/http/lua/lua_filter.cc:792] script log: 
<< response before interceptors << :status: 404
<< response before interceptors << content-length: 94
<< response before interceptors << content-type: application/json
{"code":"404","message":"Not Found","description":"The requested resource is not available."}


[2022-04-11 13:29:19.539][14][debug][lua] [source/extensions/filters/common/lua/lua.cc:39] coroutine finished
[2022-04-11 13:29:19.539][14][debug][lua] [source/extensions/filters/common/lua/lua.cc:39] coroutine finished
[2022-04-11 13:29:19.539][14][debug][lua] [source/extensions/filters/http/lua/lua_filter.cc:470] yielding for full body
[2022-04-11 13:29:19.539][14][debug][lua] [source/extensions/filters/common/lua/lua.cc:42] coroutine yielded
[2022-04-11 13:29:19.539][14][debug][lua] [source/extensions/filters/http/lua/lua_filter.cc:231] resuming body due to end stream
[2022-04-11 13:29:19.539][14][debug][lua] [source/extensions/filters/http/lua/lua_filter.cc:792] script log: 
<< response after interceptors << :status: 404
<< response after interceptors << content-length: 94
<< response after interceptors << content-type: application/json
{"code":"404","message":"Not Found","description":"The requested resource is not available."}

```
### Issues

Fixes #2360

### Automation tests
 - Unit tests added: No
 - Integration tests added: No

### Tested environments
OS: Linux
Env: Docker 

---
#### Maintainers: Check before merge
- [x] Assigned 'Type' label
- [x] Assigned the project
- [x] Validated respective github issues
- [x] Assigned milestone to the github issue(s)

### CODE REVIEW

File: adapter/pkg/config/log_types.go [hunk 1]
```diff
@@ -28,6 +28,10 @@ type accessLog struct {
 	Format  string
 }
 
+type routerDebugLog struct {
```
Comment (pubudu538) [inline adapter/pkg/config/log_types.go:31]: Shall we rename it to debugLogs? Now we have accessLogs. Also add the config to the log_config.toml

File: adapter/internal/oasparser/envoyconf/http_filters.go [hunk 2]
```diff
@@ -55,9 +53,28 @@ func getHTTPFilters() []*hcmv3.HttpFilter {
...
 		cors,
 		extAauth,
-		lua,
-		router,
 	}
+
+	logConf := config.ReadLogConfigs()
+
+	// debug log first filter
+	if logConf.DebugLogs.Enable {
```
Comment (VirajSalaka) [inline adapter/internal/oasparser/envoyconf/http_filters.go:61]: What if we define as 
if logConf.DebugLogs.Enable {
 httpFilters = []*hcmv3.HttpFilter{ cors, extAauth, getDebugLogFirstFilter(), lua, getDebugLogSecondFilter(), router}
} else {
 httpFilters = []*hcmv3.HttpFilter{ cors, extAauth, lua, router}
}

Comment (VirajSalaka) [reply adapter/internal/oasparser/envoyconf/http_filters.go:61]: IMO, it is more readable in this way.

File: resources/conf/log_config.toml [hunk 3]
```diff
@@ -29,3 +29,6 @@ logLevel = "INFO"
 enable = false
 logfile = "/tmp/envoy.access.log" # This file will be created inside router container.
 format = "[%START_TIME%] '%REQ(:METHOD)% %REQ(X-ENVOY-ORIGINAL-PATH?:PATH)% %PROTOCOL%' %RESPONSE_CODE% %RESPONSE_FLAGS% %BYTES_RECEIVED% %BYTES_SENT% %DURATION% %RESP(X-ENVOY-UPSTREAM-SERVICE-TIME)% '%REQ(X-FORWARDED-FOR)%' '%REQ(USER-AGENT)%' '%REQ(X-REQUEST-ID)%' '%REQ(:AUTHORITY)%' '%UPSTREAM_HOST%'\n"
+
+[debugLogs]
```
Comment (VirajSalaka) [inline resources/conf/log_config.toml:33]: routerDebugLogs ? wdyt

Comment (pubudu538) [reply resources/conf/log_config.toml:33]: IN the line number 26 we have a comment that states logging configs of Router. After that we have a section called accessLogs. In that case do we can go with.  debugLogs. No need to repeat router again right?

**Review by pubudu538 — APPROVED** [review]

File: adapter/internal/oasparser/envoyconf/http_filters.go [hunk 4]
```diff
@@ -44,21 +44,21 @@ import (
...
 
-	httpFilters := []*hcmv3.HttpFilter{
-		cors,
-		extAauth,
-		lua,
-		router,
+	logConf := config.ReadLogConfigs()
+
+	if logConf.DebugLogs.Enable {
+		return []*hcmv3.HttpFilter{cors, extAauth, getDebugLogFirstFilter(), lua, getDebugLogSecondFilter(), router}
```
Comment (AmaliMatharaarachchi) [inline adapter/internal/oasparser/envoyconf/http_filters.go:59]: just a query; can't we get request / response headers/body from envoy access logs or trace logs?

Comment (AmaliMatharaarachchi) [inline adapter/internal/oasparser/envoyconf/http_filters.go:59]: just a query; if somebody wants to get debug logs before and after interceptor, isn't it enough to enable trace logs? We have added adequate logs in interceptor lua scripts, haven't we.

Comment (AmaliMatharaarachchi) [reply adapter/internal/oasparser/envoyconf/http_filters.go:59]: why do we have to forcibly add debug logs, doesn't envoy provide adequate logs (just felt it's highly unlikely)?

Comment (pubudu538) [reply adapter/internal/oasparser/envoyconf/http_filters.go:59]: Envoy doesn't log the message body. In here we are looking for a log that prints headers, body, etc. Somewhat similar to wire logs.

**Comment by renuka-fernando** [comment]
Have we checked the behaviour (logs) when an interceptor is configured? does it log new headers added by the interceptor?

**Review by renuka-fernando — APPROVED** [review]

File: adapter/internal/oasparser/envoyconf/http_filters.go [hunk 5]
```diff
@@ -167,6 +167,84 @@ func getLuaFilter() *hcmv3.HttpFilter {
...
+			local response_headers = response_handle:headers()
+			local log_output = "\n"
+			for header_name, header_value in pairs(response_headers) do
+				log_output = log_output .. "<< response after interceptors << " .. header_name .. ": " .. header_value .. "\n"
+			end
+			if response_handle:body() then
+				log_output = log_output .. response_handle:body():getBytes(0, response_handle:body():length()) .. "\n"
+			end
+			response_handle:logDebug(log_output)
+		end`,
```
Comment (renuka-fernando) [inline adapter/internal/oasparser/envoyconf/http_filters.go:195]: We can remove two tab indents in this Lua script string

Comment (renuka-fernando) [reply adapter/internal/oasparser/envoyconf/http_filters.go:195]: but having these indents is not an issue, plz ignore this

File: resources/conf/log_config.toml (hunk 3 above)
Comment (VirajSalaka) [reply resources/conf/log_config.toml:33]: +1

**Comment by pubudu538** [comment]
The implementation has to be changed. Hence new PR will be sent.

**Pull request closed without merging**